src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -138,6 +138,21 @@ internal partial class Binder
                     hasErrors);
             }
 
+            if (conversion.Kind == ConversionKind.ObjectCreation)
+            {
+                var convertedObjectCreation = ConvertObjectCreationExpression(syntax, (BoundUnconvertedObjectCreationExpression)source, isCast, destination, diagnostics);
+                return new BoundConversion(
```

Comment:
>BoundConversion [](http://example.com/codeflow?start=27&length=15)

Do we really want to have this node in the tree for implicit conversions? This might have impact on SemanticModel, IOperation tree, etc. I would be cautious adding conversion nodes just for the purpose of keeping NullableWalker happy. Perhaps there is another way to indicate that object creation was target typed, adding a property to the node, etc.

Reply:
This is the same strategy used for conditionals and switch. We need the exact destination type for this to work. We already track WasTargetTyped on the bound node but it won't help when the target may have a different nullability state as demonstrated by the linked issue.

Reply:
It wasn't the initial design though. See https://github.com/dotnet/roslyn/pull/51791 where those nodes are introduced.

Reply:
I would respond the same way to that change as well. I am not sure how extra conversion node is going to be a requirement for an implicit conversion case. The target type is coming from elsewhere, NullableWalker possibly recalculates it with actual nullability information and needs to "force" it on the new conversion node. It probably can "force" it directly on the target-typed object creation node and other target typed nodes.

Reply:
I didn't realize you haven't signed off that PR. tagging @333fred for thoughts.

Reply:
The issue is that, if we don't have a single node we can force this conversion through (the BoundConversion in this case) we have to check every location in code that can provide a target type for this case.

That being said, conditional and switch expression conversions are _actually_ conversions, while this is not a conversion. It's a target-typed new, which doesn't have any kind of conversion in this location. I'm definitely leary of introducing a BoundConversion in the tree that doesn't even represent a real conversion.

Reply:
> The issue is that, if we don't have a single node we can force this conversion through (the BoundConversion in this case) we have to check every location in code that can provide a target type for this case. 

So we will have to do this for new() anyways? We will need a new bottleneck in the walker to get access to the destination without a conversion node. Messy, but I'd agree that this is not a proper "conversion" to begin with.

>That being said, conditional and switch expression conversions are actually conversions

I don't think that was the mindset initially. I do recall we treated all target-typed conversions just as a binder implementation detail that we didn't want to expose to semantic model. Is that something we no longer tend to maintain?

Reply:
The conditional and switch expression conversions are in the language spec, I think so we had a conversion that could be considered "worse" than any previously existing conversion.

Reply:
>so we had a conversion that could be considered "worse" than any previously existing conversion.

If I'm reading this right, doesn't that make *all* candidates "equally worse" in an overload resolution pass? because this is a conversion-from-expression and I don't follow how it could affect the result. Can you please elaborate a bit?

Reply:
@alrz After looking at the other similar PR and some thinking, I couldn't come up with an alternative that doesn't come with a bigger complexity. One thing I would like to suggest is to trying to mark the BoundConversion node as compiler generated in case of an implicit conversion. Would you like resuming the work on this PR with the goal to get it merged somewhere next week? Otherwise, I can do that.

Reply:
Feel free to commit to this PR or open another. There's some open issues you may want to revisit:

https://github.com/dotnet/roslyn/pull/54383#discussion_r658657515
https://github.com/dotnet/roslyn/pull/54383#discussion_r658658969
https://github.com/dotnet/roslyn/pull/54383#discussion_r658665526